relay-server/src/actors/events.rs

```diff
@@ -1387,9 +1396,112 @@ impl Handler<QueueEnvelope> for EventManager {
+    };
+
+    let sampling_config = match project_state.config.sampling_config {
+        // without sampling config we cannot sample transactions so give up here
+        None => return envelope,
+        Some(ref sampling_config) => sampling_config,
+    };
+
+    let project_id = match project_state.project_id {
+        None => return envelope,
+        Some(project_id) => project_id,
+    };
+
+    let trace_context = envelope.get_trace_context();
+    let transaction_item = envelope.get_item_by(|item| item.ty() == ItemType::Transaction);
+
+    let trace_context = match (trace_context, transaction_item) {
+        // we don't have what we need, can't sample the transactions in this envelope
+        (None, _) | (_, None) => return envelope,
+        // see if we need to sample the transaction
+        (Some(trace_context), Some(_)) => trace_context,
+    };
+
+    let should_sample = trace_context
+        // see if we should sample
+        .should_sample(sampling_config, project_id)
+        // TODO verify that this is the desired behaviour (i.e. if we can't find a rule
+        // for sampling, include the transaction)
+        .unwrap_or(true);
+
+    if !should_sample {
+        // finally we decided that we should sample the transaction
+        envelope.take_item_by(|item| item.ty() == ItemType::Transaction);
+    }
+
+    envelope
+}
+
+/// Check if we should remove transactions from this envelope (because of trace sampling) and
+/// return what is left of the envelope
+pub fn sample_transaction(
+    mut envelope: Envelope,
+    project: Option<Addr<Project>>,
+    fast_processing: bool,
+) -> ResponseFuture<Envelope, ()> {
+    let project = match project {
+        None => return Box::new(future::ok(envelope)),
+        Some(project) => project,
+    };
+
+    let trace_context = envelope.get_trace_context();
+    let transaction_item = envelope.get_item_by(|item| item.ty() == ItemType::Transaction);
+
+    // if there is no trace context or there are no transactions to sample return here
+    if trace_context.is_none() || transaction_item.is_none() {
+        return Box::new(future::ok(envelope));
+    }
+    //we have a trace_context and we have a transaction_item see if we can sample them
+    if fast_processing {
+        let fut = project.send(GetCachedProjectState).then(|project_state| {
+            let project_state = match project_state {
+                // error getting the project, give up and return envelope unchanged
+                Err(_) => return Ok(envelope),
+                Ok(project_state) => project_state,
+            };
+            sample_transaction_internal(&mut envelope, project_state.as_deref());
+            if envelope.is_empty() {
+                Err(())
+            } else {
+                Ok(envelope)
+            }
+        });
+        Box::new(fut) as ResponseFuture<_, _>
+    } else {
+        //TODO can we deduplicate the code below without using some weird macro?
```

Comment:
You can by implementing a trait, either over the message, the future, or the result that is then implemented for both versions. I don't think this is worth it, however. You could remove some duplication by moving the envelope empty check out. 

Note that the envelope may still contain items even after a transaction was sampled. I would not check for empty envelopes at this point and let the caller handle this.

Reply:
Regarding the Note:
My idea is that when we end up with empty envelopes we can shortcut the processing. 
If due to the transaction being sampled the envelope ends up Empty we terminate envelope processing and issue an outcome with the discard reason of `TransactionSampled`

Reply:
> My idea is that when we end up with empty envelopes we can shortcut the processing.

Agreed, although I think this control flow should happen in the caller, and not here by throwing an error. 

> will look into adding a Trait and if it is worthwhile.

Most likely this would just over-complicate. The duplication here is not critical, so I'm fine with it.

Reply:
Hmm, the problem is that an empty envelope that goes through check_envelope will be set to None. 
Then in analysing the response from CheckEnvelope we test against the envelope and if it is empty we raise a 
`ProcessingError::RateLimited` so the end result is that if I return an empty envelope it will end up as an RateLimited outcome. 
Of course I could change the test to check for envelope== None and rate_limits contains some rate limits and issue a ProcessingError::RateLimited and if envelope==None and rate_limits is empty issue a ProcessingError::TransactionSampled. 
This gets quite ugly ( I would issue a TransactionSampled error from the code that looks at the result of CheckEnvelope which up to now only deals with rate limits).

We can change even further (but I'm trying not to revolutionise `handle<HandleEnvelope>`) and issue a TransactionSampled outcome outside handling of ProcessingError... 

Let's look again at it together and decide if cleaning `sample_transaction` and not returning an error when the envelope is empty is worth all the changes in the calling code.

Reply:
done